src/mds/MDCache.cc

```diff
@@ -271,6 +271,9 @@ void MDCache::add_inode(CInode *in)
         g_conf->mds_cache_size * g_conf->mds_health_cache_threshold) {
     exceeded_size_limit = true;
   }
+
+  /* This is the place where inodes are thrown in the cache, so check if the inode should be somewhere else: */
+  in->maybe_export_pin();
```

Comment:
Okay thanks for pointing that out! So you're suggesting moving maybe_export_pin to `link_primary_inode`? I was looking around to see if that would cover the appropriate cases when loading an inode. I see that CDir::_load_dentry doesn't actually call into that link_primary_inode. Instead it calls add_primary_dentry which probably should also call link_primary_inode but instead does the work itself.

Should I go ahead and change add_primary_dentry to call link_primary_inode and move the maybe_export_pin call to link_primary_inode?